packages/airnode-node/src/evm/requests/blocking.ts

```diff
@@ -46,6 +45,41 @@ export function blockRequestsWithWithdrawals(requests: GroupedRequests): LogsDat
     return { ...acc, apiCalls: [...acc.apiCalls, apiCall] };
   }, initialState);
 
-  const updatedRequests = { ...requests, apiCalls };
-  return [logs, updatedRequests];
+  return [[...prevLogs, ...logs], { ...requests, apiCalls }];
+}
+
+export function applySponsorRequestLimit([prevLogs, requests]: LogsData<GroupedRequests>): LogsData<GroupedRequests> {
+  const requestCountPerSponsor = new Map<string, number>();
+  const allowedApiCalls: Request<ApiCall>[] = [];
+  const logs: PendingLog[] = [];
+
+  requests.apiCalls.forEach((apiCall) => {
+    if (apiCall.status !== RequestStatus.Pending) return;
+
+    if (!requestCountPerSponsor.has(apiCall.sponsorAddress)) {
+      requestCountPerSponsor.set(apiCall.sponsorAddress, 0);
+    }
+    const sponsorRequests = requestCountPerSponsor.get(apiCall.sponsorAddress)!;
+
+    if (sponsorRequests < MAXIMUM_SPONSOR_WALLET_REQUESTS) {
+      requestCountPerSponsor.set(apiCall.sponsorAddress, sponsorRequests + 1);
+      allowedApiCalls.push(apiCall);
+    } else {
+      logs.push(logger.pend('WARN', `Ignoring Request ID:${apiCall.id} as it exceeded sponsor wallet request limit.`));
+      const blockedCall: Request<ApiCall> = {
+        ...apiCall,
+        status: RequestStatus.Ignored,
+        errorMessage: `${RequestErrorMessage.SponsorRequestLimitExceeded}: ${apiCall.id}`,
+      };
+      allowedApiCalls.push(blockedCall);
+    }
+  });
+
+  return [[...prevLogs, ...logs], { ...requests, apiCalls: allowedApiCalls }];
 }
+
+// TODO: Merge with verification/api-call-verification.ts
+export const blockRequests = (requests: GroupedRequests): LogsData<GroupedRequests> => {
+  const filterRequests = flow(blockRequestsWithWithdrawals, applySponsorRequestLimit);
```

Comment:
Pretty neat - I didn't know about `flow`. Looks like it's similar to the proposed pipeline operator. I would love to have this available to avoid things like `state1`, `state2` etc. Hopefully one day...

Reply:
I think the time for it will come :)